src/github.com/matrix-org/dendrite/roomserver/input/latest_events.go

```diff
@@ -0,0 +1,72 @@
+package input
+
+import (
+	"bytes"
+	"github.com/matrix-org/dendrite/roomserver/types"
+	"github.com/matrix-org/gomatrixserverlib"
+)
+
+// updateLatestEvents updates the list of latest events for this room.
+func updateLatestEvents(
+	db RoomEventDatabase, roomNID types.RoomNID, stateAtEvent types.StateAtEvent, event gomatrixserverlib.Event,
+) error {
+	oldLatest, updater, err := db.GetLatestEventsForUpdate(roomNID)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err == nil {
+			// Commit if there wasn't an error.
+			// Set the returned err value if we encounter an error committing.
+			err = updater.Close(true)
+		} else {
+			// Ignore any error we get rolling back since we don't want to
+			// clobber the current error
+			updater.Close(false)
+		}
+	}()
+
+	var prevEvents []gomatrixserverlib.EventReference
+	prevEvents = event.PrevEvents()
+
+	if err = updater.StorePreviousEvents(stateAtEvent.EventNID, prevEvents); err != nil {
+		return err
+	}
+
+	// Check if this event references any of the latest events in the room.
+	var newLatest []types.StateAtEventAndReference
+	for _, l := range oldLatest {
+		for _, prevEvent := range prevEvents {
+			if l.EventID == prevEvent.EventID && bytes.Compare(l.EventSHA256, prevEvent.EventSHA256) == 0 {
+				// This event can be removed from the latest events cause we've found an event that references it.
+				continue
```

Comment:
I don't follow. We got an incoming event X and it states a bunch of `prev_events`. One of those prev events we have in our database. It's still the "latest" event right? Why do you filter it out?